### PR TITLE
Part I - Introduce ModuleUtils with ClassFinder SPI

### PR DESCRIPTION
**Pull request by sormuras**
## Overview

This PR prepares the support for selecting Java 9 modules for test discovery.

Basic idea: publish a `String`-based service providing interface for naming module(s) of interest and let an explicit module register a service implementation for actually finding classes in the module graph at runtime. This service implementation should be hosted in a different (sub-) project, like `junit-platform-commons-jpms`. See #1057 

At runtime, using Java 9, the user can add the `org.junit.platform.commons.jpms` module to the module-path and specify which module he wants to test. The `mods` directory contains the JUnit 5 artifacts (loaded as automatic modules), the `org.junit.platform.commons.jpms` module, the application modules (tested objects), and the test modules (classes with `@Test` annotated methods):

```
java
--module-path
  mods
--add-modules
  ALL-MODULE-PATH
--module
  org.junit.platform.console
--select-module
  jpms.integration
```

## New Console Launcher Options

- `--scan-module-path`
Scan all modules on the runtime boot module-path for test discovery.
See http://mail.openjdk.java.net/pipermail/jigsaw-dev/2017-September/013180.html for details.

- `-o` or `--select-module` with single argument `<module name>`
Select single module for test discovery. This option can be repeated.
Same runtime boot module-path as above, but only the test classes of the named modules are selected.

## TODO

- [ ] Add an option handle to (console) launcher that parses an list of paths like `--scan-module-path [entry1[:entry_n]]` and delegates to the `ModuleFinderSelector`
See http://download.java.net/java/jdk9/docs/api/java/lang/module/ModuleFinder.html

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-commons/src/main/java/org/junit/platform/commons/util/ModuleClassFinder.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	/**
+	 * Special name indicating a search for all classes in all modules on the module-path.
+	 */
+	String ALL_MODULE_PATH = "ALL-MODULE-PATH";
+
+	/**
+	 * Return list of classes found in the named module that may contain testable methods.
+	 *
+	 * @param moduleName name of the module to inspect or {@code ALL-MODULE-PATH}
```
Comment (marcphilipp) [inline junit-platform-commons/src/main/java/org/junit/platform/commons/util/ModuleClassFinder.java]: I think having two separate methods, e.g. `findAllClassesInModule()` and `findAllClassesInModulePath()`, would be a cleaner. Thoughts?

File: junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java [hunk 2]
```diff
@@ -150,6 +152,16 @@
...
+		// --- Java Platform Module System -------------------------------------
+
+		scanModulePath = parser.acceptsAll(asList("scan-modulepath", "scan-module-path"), //
+			"EXPERIMENTAL: Scan all modules on the module-path for test discovery.");
+
+		selectedModules = parser.acceptsAll(asList("o", "select-module"), //
+			"EXPERIMENTAL: Select single module for test discovery. This option can be repeated.") //
+				.withRequiredArg() //
+				.describedAs("module name");
+
```
Comment (marcphilipp) [inline junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java]: Should we also add support in the Vintage engine?

File: junit-platform-commons/src/main/java/org/junit/platform/commons/util/ModuleClassFinder.java (hunk 1 above)
Comment (sormuras) [reply junit-platform-commons/src/main/java/org/junit/platform/commons/util/ModuleClassFinder.java]: Having two methods makes sense here, because we have two different selectors in the launcher part. The blue-print where I took the special module name `"ALL-MODULE-PATH"` from only has a single command line option, namely: `--add-modules`. There is no `--add-all-module-path` or in our world: `--scan-module-path`.

tl;dr: Two separate methods are good - and remove the special module name `"ALL-MODULE-PATH"`.

File: junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java (hunk 2 above)
Comment (sormuras) [reply junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java]: Mh? Vintage should support both new selectors. If it doesn't then something was lost in rebasing to master.

File: junit-platform-commons/src/main/java/org/junit/platform/commons/util/ModuleUtils.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			Predicate<String> classNameFilter) {
+		Preconditions.notBlank(moduleName, "module name must not be null or blank");
+		Preconditions.notNull(classTester, "class tester must not be null");
+		Preconditions.notNull(classNameFilter, "class name filter must not be null");
+
+		ClassLoader classLoader = ClassLoaderUtils.getDefaultClassLoader();
+		List<Class<?>> classes = new ArrayList<>();
+
+		logger.config(() -> "Loading auto-detected class finders...");
+		int serviceCounter = 0;
```
Comment (sormuras) [inline junit-platform-commons/src/main/java/org/junit/platform/commons/util/ModuleUtils.java]: We could limit the single service implementation to be our own ... by comparing the class name or some other "magic ID". Or is it okay to allow third-party implementations here? Thoughts?

File: junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java (hunk 2 above)
Comment (sormuras) [reply junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java]: Find vintage-affecting changes below, around https://github.com/junit-team/junit5/pull/1061/files#diff-5e437a25c16596592cf66086916f6fed

Comment (marcphilipp) [reply junit-platform-console/src/main/java/org/junit/platform/console/options/AvailableOptions.java]: Sorry, missed that! LGTM!

**Comment by sormuras** [comment]
Superseded by #1087
